Tests/SonarScanner.MSBuild.PreProcessor.Test/Caching/FileCacheTests.cs

```diff
@@ -201,6 +200,52 @@ The calculation of the checksum of the file '{downloadTarget}' failed with messa
             """);
     }
 
+    [TestMethod]
+    public async Task DownloadAndValidateFile_Success()
+    {
+        var context = new ChecksumContext(fileCache, directoryWrapper, fileWrapper, checksum);
+
+        var result = await context.DownloadAndValidateFileTest(new MemoryStream(context.DownloadContentArray));
+
+        result.Should().BeNull();
+        context.AssertStreamDisposed();
+        fileWrapper.Received(1).Create(Path.Combine(context.DownloadPath, context.TempFileName));
+        fileWrapper.Received(1).Move(Path.Combine(context.DownloadPath, context.TempFileName), context.DownloadTarget);
+        context.FileContentArray.Should().BeEquivalentTo(context.DownloadContentArray);
+        testLogger.DebugMessages.Should().BeEquivalentTo($"The checksum of the downloaded file is '{context.ExpectedSha}' and the expected checksum is '{context.ExpectedSha}'.");
+    }
+
+    [TestMethod]
+    public async Task DownloadAndValidateFile_NullStream_ReturnsInvalidOperationException()
+    {
+        var context = new ChecksumContext(fileCache, directoryWrapper, fileWrapper, checksum);
+
+        var result = await context.DownloadAndValidateFileTest(null);
+
+        result.Should().BeOfType<InvalidOperationException>().Which.Message.Should().Be(
+            "The download stream is null. The server likely returned an error status code.");
+        context.AssertTempFileCreatedAndDeleted();
+        context.AssertStreamDisposed();
+        testLogger.DebugMessages.Should().BeEquivalentTo($"Deleting file '{Path.Combine(context.DownloadPath, context.TempFileName)}'.");
+    }
+
+    [TestMethod]
+    public async Task DownloadAndValidateFile_WrongChecksum_ReturnsCryptographicException()
+    {
+        var context = new ChecksumContext(fileCache, directoryWrapper, fileWrapper, checksum);
+        checksum.ComputeHash(context.FileContentStream).ReturnsForAnyArgs("someOtherHash");
```

Comment:
either
```suggestion
        checksum.ComputeHash(null).ReturnsForAnyArgs("someOtherHash");
```
or set up the mocking so that it triggers only on the correct parameter.